azdo: Add script to fetch/push sstate-cache

shruthi-ravi opened the pull request:
This script will be used by the azdo jobs to sync down the sstate
cache from argo at build time, as well as push the local sstate cache
to argo after the build finishes.
The script maintains a maximum of five sstate cache directories on
the remote location and will delete older directories to maintain the
the limit.

PR runs from azdo [here](https://ni.visualstudio.com/DevCentral/_git/ni-central/pullrequest/136867).
Azdo bug reference [here](https://ni.visualstudio.com/DevCentral/_workitems/edit/1299743).

@ni/rtos

shruthi-ravi committed:
azdo: Add script to fetch/push sstate-cache

This script will be used by the azdo jobs to sync down the sstate
cache from argo at build time, as well as push the local sstate cache
to argo after the build finishes.
The script maintains a maximum of five sstate cache directories on
the remote location and will delete older directories to maintain the
the limit.

Azdo Bug: 1299743

Signed-off-by: Shruthi Ravichandran <shruthi.ravichandran@ni.com>

shruthi-ravi commented on scripts/azdo/update_sstate_cache.sh:
```diff
@@ -0,0 +1,110 @@
+	mkdir $temp_argo_dir
+	echo Pushing sstate cache to argo at $temp_argo_dir
+	cp -rL $local_sstate_cache_dir/* $temp_argo_dir/
+	mv $temp_argo_dir ${temp_argo_dir%.latest}
```
Note that because this is on argo, the `mv` operation may not be atomic, and there is a possibility an incomplete cache could be consumed by build jobs. I could instead:

1. Create a temp file in the local cache directory before starting the `cp` operation, say `donotuse`
2. Remove the file in the remote directory after the `mv` operation has completed 
3. Update the `fetch_cache` operation to ignore any directory with the .latest suffix, as well a `donotuse` file

Lmk what you think.

amstewart replied:
>Update the fetch_cache operation to ignore any directory with the .latest suffix, as well a donotuse file

What about using a suffix like `*.tmp` on the argo directory while you're uploading the content and then removing that `.tmp` suffix when it is ready for consumption. I think the pull/fetch operation should avoid any directory ending in `.tmp` and simply find the latest, non-temp directory using their timestamps.

The local directory operations you talk about (`donotuse`) are unnecessary in that case.

shruthi-ravi replied:
I think the current script does what you described, albeit using a .latest suffix and not .tmp? Is there a different mechanism to remove the suffix if not with the `mv` command?
With `mv`, I was thinking that if the sequence turned out such that a new directory (with the suffix removed) gets created before the contents of the old directory are moved to the new directory, there is a chance that when fetch is called it picks up the new directory because it doesn't have the suffix, but the cache contents in the new directory would not be complete for a while.

shruthi-ravi replied:
@amstewart what do you think?

amstewart replied:
I think I got a little lost in the discussion last week. I have a more-clear picture of what you're talking about now and I think the `donotuse` poison-pill idea is a pretty elegant solution. ACK from me. :+1:

amstewart commented on scripts/azdo/update_sstate_cache.sh:
```diff
@@ -0,0 +1,110 @@
+}
+
+push_cache() {
+	if [ ! -d $local_sstate_cache_dir -o -z "$(ls -A $local_sstate_cache_dir 2>/dev/null)" ]; then
```
The `test` here will error in cases where the `$local_sstate_cache_dir` variable contains spaces. You'll want to surround it with double-quotes.

Ditto for most of the `cp` and `mkdir` operations elsewhere in the script.

amstewart commented on scripts/azdo/update_sstate_cache.sh:
```diff
@@ -0,0 +1,110 @@
+#!/bin/bash
+
```
It's probably best to add `set -eu` here, since I see a few places in the script where operations could fail (`mkdir` on the argo share for example), and we would want to bail out in those cases.

shruthi-ravi replied:
Note that I didn't add the `-u` option because I already run checks for arguments to print helpful error messages and the script starts to error out much earlier when the `-u` option is set.

amstewart commented on scripts/azdo/update_sstate_cache.sh:
```diff
@@ -0,0 +1,110 @@
+	# Copy the contents of the cache to a temp .latest directory
+	# before removing the .latest suffix from the name.
+	temp_argo_dir=$(date +"%s").latest
+	mkdir $temp_argo_dir
```
You should add a trap in the script (`TRAP function EXIT`) to clean up the `.latest` directory, if the script fails during the upload.

amstewart commented on scripts/azdo/update_sstate_cache.sh:
```diff
@@ -0,0 +1,110 @@
+
+$(basename $BASH_SOURCE) fetch local_sstate_cache_dir argo_sstate_cache_dir
+   Fetch the latest sstate cache from the remote argo cache directory
+   and install to the local sstate-cache directory.
```
MINOR: I think the commands should be `push` and `pull` (rather than `fetch`), in keeping with `git` terminology.

amstewart commented on scripts/azdo/update_sstate_cache.sh:
```diff
@@ -0,0 +1,110 @@
+	cp -rL $local_sstate_cache_dir/* $temp_argo_dir/
+	mv $temp_argo_dir ${temp_argo_dir%.latest}
+
+	# Cap the number of cache directories on the argo server at
+	# max_dir_count. Delete older cache directories exceeding the
+	# cap.
```
As this block is written, it just deletes any content in argo. It should discriminate between the temp/latest directories and the _complete_ directories.

Also, I think you're relying on `find` to sort the directories, and I don't think that's reliable. I'm not sure it will always delete the oldest timestamp.

shruthi-ravi replied:
There shouldn't be any .latest directories at this point. A .latest directory if created before would have already been renamed above.
I'm relying on `find` to count the number of directories before proceeding to delete. I could have used `ls`, but I didn't think of it then and don't think it will change the result now. I use `ls` later on to sort (by file modification time, not the name of the directory) and pick the directories to remove.

amstewart replied:
>There shouldn't be any .latest directories at this point.

What if two pushes are being run concurrently?

>I use ls later on to sort (by file modification time, not the name of the directory) and pick the directories to remove.

Hmm.. I'm not sure we can rely on file modification time across the CIFS boundary. Since they're already named using UNIX-time, why not rely on that?

shruthi-ravi replied:
> What if two pushes are being run concurrently?

Only the CI jobs will attempt pushes, and they are scheduled to run sequentially.

>Hmm.. I'm not sure we can rely on file modification time across the CIFS boundary. Since they're already named using UNIX-time, why not rely on that?

Sure.

amstewart replied:
>Only the CI jobs will attempt pushes, and they are scheduled to run sequentially.

I agree that *shouldn't* happen, but I'm uncomfortable relying on that assumption. I'll defer to you as to whether or not the enhancement is worth it.

shruthi-ravi committed:
Miscellaneous updates and bug fixes.

- Add `set -e`
- Replace `fetch` with `pull`
- Use quotes when referring to the argo cache and local sstate cache
  dirs
- Add a trap to delete the .latest directory if things go wrong
- Create a DONOTUSE file in the remote cache directory location
  before copying the cache in a push operation. This is to safeguard
  against pulling the cache directory before the mv operation can be
  completed, which could result in an incomplete local cache.
  The pull operation will now only fetch the latest directory from
  argo that does not have the .latest suffix as well as the DONOTUSE
  file.
- Sort and remove older cache directories on argo based on the name
  of the directories instead of file modification time.

Signed-off-by: Shruthi Ravichandran <shruthi.ravichandran@ni.com>

shruthi-ravi committed:
Add note about sequential push operations assumption.

Signed-off-by: Shruthi Ravichandran <shruthi.ravichandran@ni.com>

shruthi-ravi commented:
**Updates**
v2: 
- Add `set -e`
- Replace `fetch` with `pull`
- Use quotes when referring to the argo cache and local sstate cache
  dirs
- Add a trap to delete the .latest directory if things go wrong
- Create a DONOTUSE file in the remote cache directory location
  before copying the cache in a push operation. This is to safeguard
  against pulling the cache directory before the mv operation can be
  completed, which could result in an incomplete local cache.
  The pull operation will now only fetch the latest directory from
  argo that does not have the .latest suffix as well as the DONOTUSE
  file.
- Sort and remove older cache directories on argo based on the name
  of the directories instead of file modification time.

v3:
- Add note about the sequential push operations assumption this script
  makes. There are several changes required to support parallel push 
  operations. To keep it simple and support the current use case where
  CI builds are run sequentially, stick to that assumption.

amstewart commented on scripts/azdo/update_sstate_cache.sh:
```diff
@@ -0,0 +1,125 @@
+	# atomic on the argo share.
+
+	temp_argo_dir=$(date +"%s").latest
+	trap "{ rm -rf $temp_argo_dir; }" EXIT
```
`TRAP`s have scope until the process exits, so this `TRAP` will be in place for the entirety of this script. It's probably inconsequential - since `$temp_argo_dir` should not exist for very long anyway - but it would be cleaner if you unset the TRAP after the target directory gets renamed (line 49). You can use the provisioning test script I wrote recently [as a reference](https://github.com/ni/nilrt/blob/nilrt/master/dunfell/scripts/tests/provisioning/start-vm-using-expect.sh#L126) for how to do that.

amstewart reviewed (changes requested):
lgtm; apart from the TRAP sticking around for longer than it should.

shruthi-ravi committed:
Unset trap after the target directory is renamed.

Signed-off-by: Shruthi Ravichandran <shruthi.ravichandran@ni.com>

amstewart reviewed (approved):
ACK. Looks good to me. :+1:

shruthi-ravi commented:
Ugh, gross! I wanted to squash my commits before rebasing. Can't fix it now can I? :( @amstewart

shruthi-ravi commented:
I'm probably going to stick to force pushes..

amstewart commented:
I don't think it's worth fixing. The commit requirements for `nilrt.git` are relatively loose, since it has no other upstream and doesn't get rebased.
